cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CQueue.scala

```diff
@@ -0,0 +1,142 @@
+package com.twitter.cassovary.util.collections
+
+import it.unimi.dsi.fastutil.PriorityQueue
+import it.unimi.dsi.fastutil.ints.{IntArrayFIFOQueue, IntComparator, IntHeapPriorityQueue,
+  IntPriorityQueue}
+import it.unimi.dsi.fastutil.objects.{ObjectArrayFIFOQueue, ObjectArrayPriorityQueue}
+
+/**
+ * Specialized implementation of a queue. Queue allows adding elements and
+ * removing them in a predefined order (defined in subclasses).
+ *
+ * Implemented methods are not boxed, when using with some `Int` and `Long`.
+ */
+trait CQueue[@specialized(Int, Long) T] {
+  /**
+   * @return Next element to be dequeued.
+   */
+  def first(): T
+
+  /**
+   *
+   * @return Next element that is removed from the `CQueue`.
+   */
+  def deque(): T
+
+  /**
+   * Adds element to the `CQueue`.
+   */
+  def +=(elem: T)
+
+  def isEmpty: Boolean
+  def clear(): Unit
+  def size(): Int
+}
+
+/**
+ * FIFO [[CQueue]] - elements are added to the end of an underlying list and
+ * removed from the beginning.
+ */
+trait FIFO[@specialized(Int, Long) T] extends CQueue[T] {
+  /**
+   * Adds an element at the beginning of the underlying list.
+   */
+  def enqueueFirst(elem: T)
+}
+
+object CQueue {
+  import Implicits._
```

Comment:
We need the traits for implicits precedense. Basically: if you can CQueue[Int]() you are matching both factories: CQueueFactory[T] called factoryAny from LowerPriorityImplicit and CQueueFactory[Int] called factoryInt from LowPriorityImplicit. 
Scala will choose the latter one, because it's defined in the subclass of the class where the former is defined. Thus, I'm afraid we can't get rid of this nasty traits...
Alternatively, we can make CQueue extends LowPriorityImplicits but this seems not intuitive for me.